src/writer.rs

```diff
@@ -0,0 +1,492 @@
+    n_addressed_tiles: u64,
+    // TODO: Replace with digest HashMap for deduplicating non-subsequent tiles
+    n_tile_contents: u64,
+    prev_tile_data: Vec<u8>,
+}
+
+pub(crate) trait WriteTo {
+    fn write_to<W: Write>(&self, writer: &mut W) -> std::io::Result<()>;
+
+    fn write_compressed_to<W: Write>(
+        &self,
+        writer: &mut W,
+        compression: Compression,
+    ) -> PmtResult<()> {
+        match compression {
+            Compression::None => self.write_to(writer)?,
+            Compression::Gzip => {
+                let mut encoder = GzEncoder::new(writer, flate2::Compression::default());
+                self.write_to(&mut encoder)?;
+            }
+            v => Err(UnsupportedCompression(v))?,
+        }
+        Ok(())
+    }
+
+    fn write_compressed_to_counted<W: Write>(
+        &self,
+        writer: &mut Counter<W>,
+        compression: Compression,
+    ) -> PmtResult<usize> {
+        let pos = writer.writer_bytes();
+        self.write_compressed_to(writer, compression)?;
+        Ok(writer.writer_bytes() - pos)
+    }
+
+    fn compressed_size(&self, compression: Compression) -> PmtResult<usize> {
+        let mut devnull = Counter::new(std::io::sink());
+        self.write_compressed_to(&mut devnull, compression)?;
+        Ok(devnull.writer_bytes())
+    }
+}
+
+impl WriteTo for [u8] {
+    fn write_to<W: Write>(&self, writer: &mut W) -> std::io::Result<()> {
+        writer.write_all(self)
+    }
+}
+
+impl PmTilesWriter {
+    /// Create a new `PMTiles` writer with default values.
+    #[must_use]
+    pub fn new(tile_type: TileType) -> Self {
+        let tile_compression = match tile_type {
+            TileType::Mvt => Compression::Gzip,
+            _ => Compression::None,
+        };
+        #[allow(clippy::excessive_precision, clippy::unreadable_literal)]
+        let header = Header {
+            version: 3,
+            root_offset: HEADER_SIZE as u64,
+            root_length: 0,
+            metadata_offset: MAX_INITIAL_BYTES as u64,
+            metadata_length: 0,
+            leaf_offset: 0,
+            leaf_length: 0,
+            data_offset: 0,
+            data_length: 0,
+            n_addressed_tiles: None,
+            n_tile_entries: None,
+            n_tile_contents: None,
+            clustered: true,
+            internal_compression: Compression::Gzip,
+            tile_compression,
+            tile_type,
+            min_zoom: 0,
+            max_zoom: 22,
+            min_longitude: -180.0,
+            min_latitude: -85.051129,
+            max_longitude: 180.0,
+            max_latitude: 85.051129,
+            center_zoom: 0,
+            center_longitude: 0.0,
+            center_latitude: 0.0,
+        };
+        Self {
+            header,
+            metadata: "{}".to_string(),
+        }
+    }
+
+    /// Set the compression for metadata and directories.
+    #[must_use]
+    pub fn internal_compression(mut self, compression: Compression) -> Self {
```

Comment:
`PmTilesWriter` *is* a builder struct, generating a `PmTilesStreamWriter` struct with `create`. `Builder` is not in the name, because the resulting code looks better without it, IMO.

Reply:
If we switched to `pmtiles::writer::StreamWriter` then we could have `pmtiles::writer::Builder` without confusion.

Reply:
I _think_ we can accomplish the same thing with name spacing: `pmtiles::writer::StreamBuilder` and `pmtiles::writer::StreamWriter`. It's a pretty common pattern in Rust, and I think keeps things concise and manageable.

Any thoughts on this @nyurik?

Reply:
It took me a few seconds to understand "name spacing" meant "namespacing" :D  Naming is hard...

I am not a big fan of mega-generic names like `Builder` by itself, nor adding namespace as part of the required pattern... I do like `PmTilesWriter::new(...).create(...)` as it is cleaner to the user.  We could name the resulting struct as `PmTilesWriterStream` or `PmTilesStream` or whatever.

Reply:
Okay, this is a pretty fundamental decision—so I'd say this blocks the MR until we decide.

I did some research, and here's my suggestion.

The "thing" we want to build is called `PmTilesWriter` (I think "stream" is an implementation—99% of use-cases prefer streaming anyways.).

Then we can have a `Builder` in the same module, but hide the complexity behind a `PmTilesWriter::builder()` call.

This results in the following pseudocode:

```rust
let writer = PmTilesWriter::builder(tile_type).create("file.pmtiles");
```

The user never needs to know about the underlying builder structure, and it keeps the naming manageable. I'm happy to do the refactor, but just want to make sure there's no strong objections.